File: Android/PRApp/app/src/main/java/com/prapp/model/db/wrapper/WEvento.java
/*
 * PRApp  Copyright (C) 2019  Luca Bartolomei
 *
 * This file is part of PRApp.
 *
 *     PRApp is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     PRApp is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with PRApp.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.prapp.model.db.wrapper;

import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;
import com.prapp.model.db.enums.StatoEvento;
import com.prapp.model.net.Twinned;
import com.prapp.model.net.serialize.JSONDeserializable;
import com.prapp.model.net.serialize.JSONSerializable;
import com.prapp.model.net.serialize.adapter.DateTimeAdapter;
import com.prapp.model.net.serialize.adapter.StatoEventoAdapter;

import org.joda.time.DateTime;

public class WEvento implements  DatabaseWrapper, Twinned, JSONSerializable, JSONDeserializable {

    public static final String CLASS_PATH = "com\\model\\db\\wrapper\\WEvento";

    public static WEvento getEmpty()
    {
        return new WEvento(0,0,0,"","","1970-01-01T00:00:00.000Z","1970-01-01T00:00:00.000Z","",StatoEvento.VALIDO, 0,"1970-01-01T00:00:00.000Z");
    }

    @SerializedName("id")
    private Integer id;

    @SerializedName("idStaff")
    private Integer idStaff;

    @SerializedName("idCreatore")
    private Integer idCreatore;

    @SerializedName("nome")
    private String nome;

    //Optional
    @SerializedName("descrizione")
    private String descrizione;

    @SerializedName("inizio")
    @JsonAdapter(DateTimeAdapter.class)
    private DateTime inizio;

    @SerializedName("fine")
    @JsonAdapter(DateTimeAdapter.class)
    private DateTime fine;

    @SerializedName("indirizzo")
    private String indirizzo;

    /*private String città;
    private String provincia;
    private String stato;*/

    @SerializedName("statoEvento")
    @JsonAdapter(StatoEventoAdapter.class)
    private StatoEvento statoEvento;

    @SerializedName("idModificatore")
    private Integer idModificatore;

    @SerializedName("timestampUltimaModifica")
    @JsonAdapter(DateTimeAdapter.class)
    private DateTime timestampUltimaModifica;


    public WEvento(Integer id, Integer idStaff, Integer idCreatore, String nome, String descrizione, DateTime inizio, DateTime fine, String indirizzo, StatoEvento statoEvento, Integer idModificatore, DateTime timestampUltimaModifica) {
        this.id = id;
        this.idStaff = idStaff;
        this.idCreatore = idCreatore;
        this.nome = nome;
        this.descrizione = descrizione;
        this.inizio = inizio;
        this.fine = fine;
        this.indirizzo = indirizzo;
        this.statoEvento = statoEvento;
        this.idModificatore = idModificatore;
        this.timestampUltimaModifica = timestampUltimaModifica;
    }

    public WEvento(Integer id, Integer idStaff, Integer idCreatore, String nome, String descrizione, String inizio, String fine, String indirizzo, StatoEvento statoEvento, Integer idModificatore, String timestampUltimaModifica) {
        this(id, idStaff, idCreatore, nome, descrizione, new DateTime(timestampUltimaModifica), new DateTime(timestampUltimaModifica), indirizzo, statoEvento, idModificatore, new DateTime(timestampUltimaModifica));
    }

    public Integer getId() {
        return id;
    }

    public Integer getIdStaff() {
        return idStaff;
    }

    public Integer getIdCreatore() {
        return idCreatore;
    }

    public String getNome() {
        return nome;
    }

    public String getDescrizione() {
        return descrizione;
    }

    public DateTime getInizio() {
        return inizio;
    }

    public DateTime getFine() {
        return fine;
    }

    public String getIndirizzo() {
        return indirizzo;
    }

    public StatoEvento getStatoEvento() {
        return statoEvento;
    }

    public Integer getIdModificatore() {
        return idModificatore;
    }

    public DateTime getTimestampUltimaModifica() {
        return timestampUltimaModifica;
    }

    public boolean isDescrizionePresent()
    {
        return descrizione != null;
    }


    @Override
    public String getRemoteClassPath() {
        return CLASS_PATH;
    }

}

File: Server/www/framework/com/model/db/wrapper/WEvento.php
<?php

/*
 * PRApp  Copyright (C) 2019  Luca Bartolomei
 *
 * This file is part of PRApp.
 *
 *     PRApp is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     PRApp is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with PRApp.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace com\model\db\wrapper;

use InvalidArgumentException;
use com\model\db\enum\StatoEvento;
use ReflectionClass;
use com\model\net\serialize\ArrayDeserializable;
use DateTimeImmutable;
use com\utils\DateTimeImmutableAdapterJSON;

class WEvento implements DatabaseWrapper
{

    const NOME_MAX = 150;

    const INDIRIZZO_MAX = 150;

    const DESCRIZIONE_MAX = 500;

    /**
     * Converte un array di stringhe in un wrapper.
     *
     * @param array $array
     * @throws InvalidArgumentException i dati dell'array non sono validi oppure l'array stesso non è valido
     * @throws \com\model\exception\ParseException
     * @return WUtente wrapper convertito
     */
    public static function of($array)
    {
        if (is_null($array) || ! is_array($array))
            throw new InvalidArgumentException("Array nullo o non valido.");

        if (! array_key_exists("id", $array))
            throw new InvalidArgumentException("Dato id non trovato.");

        if (! array_key_exists("idStaff", $array))
            throw new InvalidArgumentException("Dato idStaff non trovato.");

        if (! array_key_exists("idCreatore", $array))
            throw new InvalidArgumentException("Dato idCreatore non trovato.");

        if (! array_key_exists("nome", $array))
            throw new InvalidArgumentException("Dato nome non trovato.");

        if (! array_key_exists("descrizione", $array))
			$array["descrizione"] = null;
            //throw new InvalidArgumentException("Dato descrizione non trovato.");

        if (! array_key_exists("inizio", $array))
            throw new InvalidArgumentException("Dato inizio non trovato.");

        if (! array_key_exists("fine", $array))
            throw new InvalidArgumentException("Dato fine non trovato.");

        if (! array_key_exists("indirizzo", $array))
            throw new InvalidArgumentException("Dato indirizzo non trovato.");

        if (! array_key_exists("stato", $array))
            throw new InvalidArgumentException("Dato stato non trovato.");

        if (! array_key_exists("idModificatore", $array))
            $array["idModificatore"] = null;
            //throw new InvalidArgumentException("Dato idModificatore non trovato.");

        if (! array_key_exists("timestampUltimaModifica", $array))
            throw new InvalidArgumentException("Dato timestampUltimaModifica non trovato.");

        
        $inizio = new DateTimeImmutableAdapterJSON(\DateTimeImmutable::createFromFormat(DateTimeImmutableAdapterJSON::MYSQL_TIMESTAMP, $array["inizio"]));                                      //MYSQL TIMESTAMP
        $fine = new DateTimeImmutableAdapterJSON(\DateTimeImmutable::createFromFormat(DateTimeImmutableAdapterJSON::MYSQL_TIMESTAMP, $array["fine"]));                                          //MYSQL TIMESTAMP
        $timestampUltimaModifica = new DateTimeImmutableAdapterJSON(\DateTimeImmutable::createFromFormat(DateTimeImmutableAdapterJSON::MYSQL_TIMESTAMP, $array["timestampUltimaModifica"]));    //MYSQL TIMESTAMP
        $stato = StatoEvento::parse($array["stato"]);                                                   //INT VALUE
        $idModificatore = is_null($array["idModificatore"]) ? NULL : ((int) $array["idModificatore"]);  //NULL PARSE


        return self::make((int) $array["id"], (int) $array["idStaff"], (int) $array["idCreatore"], $array["nome"], $array["descrizione"], $inizio, $fine, $array["indirizzo"], $stato, $idModificatore, $timestampUltimaModifica);
    }

    /**
     * Metodo factory con controlli.
     *
     * @param int $id
     * @param int $idStaff
     * @param int $idCreatore
     * @param string $nome
     * @param string $descrizione
     * @param \DateTimeImmutableAdapterJSON $inizio
     * @param \DateTimeImmutableAdapterJSON $fine
     * @param string $indirizzo
     * @param string $città
     * @param string $provincia
     * @param string $stato
     * @param StatoEvento $statoEvento
     * @param int $idModificatore
     * @param \DateTimeImmutableAdapterJSON $timestampUltimaModifica
     * @throws InvalidArgumentException
     * @return \com\model\db\wrapper\WEvento
     */
    public static function make($id, $idStaff, $idCreatore, $nome, $descrizione, $inizio, $fine, $indirizzo, $stato, $idModificatore, $timestampUltimaModifica)
    {
        if (is_null($id) || is_null($idStaff) || is_null($idCreatore) || is_null($nome) || is_null($inizio) || is_null($fine) || is_null($indirizzo) || is_null($stato) || is_null($timestampUltimaModifica))
            throw new InvalidArgumentException("Uno o più parametri nulli");

        if (! is_int($id) || ! is_int($idStaff) || ! is_int($idCreatore) || ! is_string($nome) || (! is_null($descrizione) && ! is_string($descrizione)) || ! ($inizio instanceof DateTimeImmutableAdapterJSON) || ! ($fine instanceof DateTimeImmutableAdapterJSON) || ! is_string($indirizzo) || ! ($stato instanceof StatoEvento) || ! ($timestampUltimaModifica instanceof DateTimeImmutableAdapterJSON) || (! is_null($idModificatore) && ! is_int($idModificatore)))
            throw new InvalidArgumentException("Uno o più parametri non del tipo giusto");

        if (strlen($nome) > self::NOME_MAX)
            throw new InvalidArgumentException("Nome non valido (MAX)");

        if (strlen($indirizzo) > self::INDIRIZZO_MAX)
            throw new InvalidArgumentException("Indirizzo non valido (MAX)");

        if ($id <= 0)
            throw new InvalidArgumentException("ID non valido");

        if ($idStaff <= 0)
            throw new InvalidArgumentException("ID Staff non valido");

        if ($idCreatore <= 0)
            throw new InvalidArgumentException("ID Creatore non valido");

        if (! is_null($idModificatore) && $idModificatore <= 0)
            throw new InvalidArgumentException("ID Modificatore non valido");

        if ($inizio->getDateTimeImmutable() > $fine->getDateTimeImmutable())
            throw new InvalidArgumentException("Ordine cronologico date non valido");

        if (!is_null($descrizione)){
			if(strlen($descrizione) > self::DESCRIZIONE_MAX)
				throw new InvalidArgumentException("Descrizione non valida (MAX)");
		}
		

        return new WEvento($id, $idStaff, $idCreatore, $nome, $descrizione, $inizio, $fine, $indirizzo, $stato, $idModificatore, $timestampUltimaModifica);
    }

    public static function makeNoChecks($id, $idStaff, $idCreatore, $nome, $descrizione, $inizio, $fine, $indirizzo, $stato, $idModificatore, $timestampUltimaModifica)
    {
        return new WEvento($id, $idStaff, $idCreatore, $nome, $descrizione, $inizio, $fine, $indirizzo, $stato, $idModificatore, $timestampUltimaModifica);
    }

    /**
     * Identificativo dell'evento.
     * (>0).
     *
     * @var int|NULL
     */
    private $id;

    /**
     * Identificativo dello staff che ha creato l'evento.
     * (>0).
     *
     * @var int|NULL
     */
    private $idStaff;

    /**
     * Identificativo dell'utente che ha creato l'evento.
     *
     * @var int|NULL
     */
    private $idCreatore;

    /**
     * Nome dell'evento.
     *
     * @var string|NULL
     */
    private $nome;

    /**
     * Descrizione dell'evento.
     * (OPZIONALE).
     *
     * @var string
     */
    private $descrizione;

    /**
     * Inizio dell'evento.
     *
     * @var DateTimeImmutableAdapterJSON
     */
    private $inizio;

    /**
     * Fine dell'evento.
     *
     * @var DateTimeImmutableAdapterJSON
     */
    private $fine;

    /**
     * Indirizzo dove si terrà l'evento.
     *
     * @var string
     */
    private $indirizzo;

    /**
     * Stato dell'evento.
     *
     * @var StatoEvento
     */
    private $stato;

    /**
     * Identificativo dell'ultimo amministratore che ha modificato l'evento.
     * (>0).
     *
     * @var int|NULL
     */
    private $idModificatore;

    /**
     * Indica quando è stata effettuata l'ultima modifica.
     *
     * @var DateTimeImmutableAdapterJSON
     */
    private $timestampUltimaModifica;

    protected function __construct($id, $idStaff, $idCreatore, $nome, $descrizione, $inizio, $fine, $indirizzo, $stato, $idModificatore, $timestampUltimaModifica)
    {
        $this->id = $id;
        $this->idStaff = $idStaff;
        $this->idCreatore = $idCreatore;
        $this->nome = $nome;
        $this->descrizione = $descrizione;
        $this->inizio = $inizio;
        $this->fine = $fine;
        $this->indirizzo = $indirizzo;
        $this->stato = $stato;
        $this->idModificatore = $idModificatore;
        $this->timestampUltimaModifica = $timestampUltimaModifica;
    }

    /**
     * Serve per serializzare il wrapper in JSON.
     */
    public function jsonSerialize()
    {
        $thisClass = new ReflectionClass(get_class());
        $props = $thisClass->getProperties();
        $array = array();

        foreach ($props as $prop) {
            $array[$prop->getName()] = $this->{$prop->getName()};
        }

        return $array;
    }

    /**
     *
     * @return number|NULL
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     *
     * @return number
     */
    public function getIdStaff()
    {
        return $this->idStaff;
    }

    /**
     *
     * @return number|NULL
     */
    public function getIdCreatore()
    {
        return $this->idCreatore;
    }

    /**
     *
     * @return string
     */
    public function getNome()
    {
        return $this->nome;
    }

    /**
     *
     * @return string|bool
     */
    public function getDescrizione()
    {
        return is_null($this->descrizione) ? FALSE : $this->descrizione;
    }

    /**
     *
     * @return DateTimeImmutableAdapterJSON
     */
    public function getInizio()
    {
        return $this->inizio;
    }

    /**
     *
     * @return DateTimeImmutableAdapterJSON
     */
    public function getFine()
    {
        return $this->fine;
    }

    /**
     *
     * @return string
     */
    public function getIndirizzo()
    {
        return $this->indirizzo;
    }

    /**
     *
     * @return \com\model\db\enum\StatoEvento
     */
    public function getStato()
    {
        return $this->stato;
    }

    /**
     *
     * @return number|NULL
     */
    public function getIdModificatore()
    {
        return $this->idModificatore;
    }

    /**
     *
     * @return DateTimeImmutableAdapterJSON
     */
    public function getTimestampUltimaModifica()
    {
        return $this->timestampUltimaModifica;
    }

}

